src/Symfony/Component/DependencyInjection/ContainerBuilder.php

```diff
@@ -348,21 +343,11 @@ public function set($id, $service)
 
         if ($this->isFrozen()) {
             // setting a synthetic service on a frozen container is alright
-            if (
-                (!isset($this->definitions[$id]) && !isset($this->obsoleteDefinitions[$id]))
-                    ||
-                (isset($this->definitions[$id]) && !$this->definitions[$id]->isSynthetic())
-                    ||
-                (isset($this->obsoleteDefinitions[$id]) && !$this->obsoleteDefinitions[$id]->isSynthetic())
-            ) {
+            if (!isset($this->definitions[$id]) || !$this->definitions[$id]->isSynthetic()) {
                 throw new BadMethodCallException(sprintf('Setting service "%s" on a frozen container is not allowed.', $id));
```

Comment:
What about a check if `$service` is an instance of `$definition->getClass()` for sanity. Or "synchronize" the definition here accordingly, ie `$definition->setClass(get_class($service));`